In [1]:
# Import libraries to be used
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
import warnings
warnings.filterwarnings('ignore')
from pycaret.classification import *


In [2]:
# Import finalized data which was cleaned and transformed in the previous notebook
df = pd.read_csv('credit_risk_transformed_final.csv')
df.head()

Id       Age     Income  Emp_length     Amount   Rate  Status  \
0   0  3.135494  10.985310    1.526056  10.463132  16.02       1   
1   1  3.091042   9.169623    1.791759   6.908755  11.14       0   
2   2  3.258097   9.169623    0.693147   8.612685  12.87       1   
3   3  3.178054  11.089821    1.609438  10.463132  15.23       1   
4   4  3.218876  10.904138    2.197225  10.463132  14.27       1   

   Percent_income  Default  Cred_length  Home_MORTGAGE  Home_OTHER  Home_OWN  \
0            0.59        1            3          False       False     False   
1            0.10        0            2          False       False      True   
2            0.57        0            3           True       False     False   
3            0.53        0            2          False       False     False   
4            0.55        1            4          False       False     False   

   Home_RENT  Intent_DEBTCONSOLIDATION  Intent_EDUCATION  \
0       True                     False             False   
1      False                     False              True   
2      False                     False             False   
3       True                     False             False   
4       True                     False             False   

   Intent_HOMEIMPROVEMENT  Intent_MEDICAL  Intent_PERSONAL  Intent_VENTURE  
0                   False           False             True           False  
1                   False           False            False           False  
2                   False            True            False           False  
3                   False            True            False           False  
4                   False            True            False           False

In [3]:
# Setup the environment and preprocess the data
clf1 = setup(data = df, target = 'Status', use_gpu=True, session_id=123)

[LightGBM] [Warning] There are no meaningful features which satisfy the provided configuration. Decreasing Dataset parameters min_data_in_bin or min_data_in_leaf and re-constructing Dataset might resolve this warning.
[LightGBM] [Info] Number of positive: 1, number of negative: 1
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 0
[LightGBM] [Info] Number of data points in the train set: 2, number of used features: 0
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 3080 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 16 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Warning] GPU acceleration is disabled because no non-trivial dense features can be found
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.500000 -> initscore=0.000000
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that 

[LightGBM] [Warning] There are no meaningful features which satisfy the provided configuration. Decreasing Dataset parameters min_data_in_bin or min_data_in_leaf and re-constructing Dataset might resolve this warning.
[LightGBM] [Info] Number of positive: 1, number of negative: 1
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 0
[LightGBM] [Info] Number of data points in the train set: 2, number of used features: 0
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 3080 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 16 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Warning] GPU acceleration is disabled because no non-trivial dense features can be found
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.500000 -> initscore=0.000000
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[LightGBM] [Warning] Stopped training because there are no more leaves that 

In [4]:
# Compare baseline models and select candidate model or models for further tuning
best_model = compare_models()

Processing:   0%|          | 0/69 [00:00<?, ?it/s]

In [5]:
# Evaluate the results of model comparison
train_results = clf1.pull()
train_results

Model  Accuracy     AUC  Recall   Prec.  \
xgboost         Extreme Gradient Boosting    0.9320  0.9534  0.7357  0.9395   
lightgbm  Light Gradient Boosting Machine    0.9271  0.9460  0.6961  0.9586   
rf               Random Forest Classifier    0.9180  0.9256  0.6691  0.9368   
catboost              CatBoost Classifier    0.9178  0.9304  0.6647  0.9411   
gbc          Gradient Boosting Classifier    0.9137  0.9196  0.6517  0.9322   
et                 Extra Trees Classifier    0.9031  0.9058  0.6223  0.9039   
dt               Decision Tree Classifier    0.8777  0.8255  0.7331  0.7144   
ada                  Ada Boost Classifier    0.8756  0.8881  0.6046  0.7756   
lda          Linear Discriminant Analysis    0.8476  0.8545  0.4762  0.7318   
ridge                    Ridge Classifier    0.8443  0.0000  0.3911  0.7883   
lr                    Logistic Regression    0.8170  0.7838  0.2967  0.6761   
nb                            Naive Bayes    0.8117  0.8188  0.4665  0.5862   
dummy                    Dummy Classifier    0.7819  0.5000  0.0000  0.0000   
knn                K Neighbors Classifier    0.7812  0.6648  0.2378  0.4971   
qda       Quadratic Discriminant Analysis    0.7771  0.8039  0.6579  0.5069   
svm                   SVM - Linear Kernel    0.6165  0.0000  0.3426  0.5219   

              F1   Kappa     MCC  TT (Sec)  
xgboost   0.8251  0.7837  0.7928     0.656  
lightgbm  0.8065  0.7630  0.7778     0.655  
rf        0.7806  0.7319  0.7476     0.832  
catboost  0.7791  0.7304  0.7471     0.624  
gbc       0.7670  0.7161  0.7334     0.705  
et        0.7370  0.6801  0.6980     0.871  
dt        0.7234  0.6449  0.6451     0.730  
ada       0.6794  0.6037  0.6110     0.705  
lda       0.5768  0.4889  0.5058     0.660  
ridge     0.5226  0.4420  0.4809     0.634  
lr        0.4029  0.3186  0.3584     0.673  
nb        0.5193  0.4042  0.4084     0.649  
dummy     0.0000  0.0000  0.0000     0.614  
knn       0.3215  0.2101  0.2305     0.766  
qda       0.5644  0.4211  0.4331     0.661  
svm       0.1839  0.0519  0.0977     0.626

In [6]:
# Save results for evaluation later if needed
# train_results.to_csv('PyCaret/train_results.csv')

In [7]:
print(best_model)

XGBClassifier(base_score=None, booster='gbtree', callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, gpu_id=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              n_estimators=100, n_jobs=-1, num_parallel_tree=None,
              objective='binary:logistic', predictor=None, ...)


In [8]:
evaluate_model(best_model)

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Pipeline Plot', 'pipelin…

In [9]:
tuned_best_model = tune_model(best_model, n_iter=50, early_stopping=True, early_stopping_max_iters=10)


Processing:   0%|          | 0/7 [00:00<?, ?it/s]

Fitting 10 folds for each of 50 candidates, totalling 500 fits
Original model was better than the tuned model, hence it will be returned. NOTE: The display metrics are for the tuned model (not the original one).


In [10]:
print(tuned_best_model)

XGBClassifier(base_score=None, booster='gbtree', callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, gpu_id=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=None, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=None, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              n_estimators=100, n_jobs=-1, num_parallel_tree=None,
              objective='binary:logistic', predictor=None, ...)


In [11]:
prediction_model = predict_model(tuned_best_model)

In [12]:
evaluate_model(tuned_best_model)

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Pipeline Plot', 'pipelin…

In [13]:
from pycaret.classification import plot_model

def plot_all_model_plots(model, scale=1, save=True, fold=None, fit_kwargs=None, plot_kwargs=None, groups=None, verbose=True, display_format='png'):
    """
    Plot all available plots for a given trained classification model using PyCaret's plot_model function.

    Parameters:
    - model: scikit-learn compatible object
        The trained model object.
    - scale: float, default=1
        Scale factor for the plots.
    - save: bool, default=False
        Whether to save the plots as image files.
    - fold: int or None, optional
        The fold number for cross-validation plots. If None, the default fold (fold 0) is used.
    - fit_kwargs: dict or None, optional
        Additional keyword arguments for the fit function (if retraining is required).
    - plot_kwargs: dict or None, optional
        Additional keyword arguments for plot customization.
    - groups: str or None, optional
        Group variable for plotting in case of group-based plots.
    - verbose: bool, default=True
        Whether to display the plot names as they are generated.
    - display_format: str or None, optional
        Format for displaying the plots (e.g., 'png', 'pdf').

    Returns:
    - None
    """

    # List of available plots
    available_plots = [
        'pipeline',
        'auc',
        'threshold',
        'pr',
        'confusion_matrix',
        'error',
        'class_report',
        'boundary',
        'rfe',
        'learning',
        'manifold',
        'calibration',
        'vc',
        'dimension',
        'feature',
        'feature_all',
        'parameter',
        'lift',
        'gain',
        'tree',
        'ks'
    ]

    for plot_type in available_plots:
        try:
            plot_model(model, plot=plot_type, scale=scale, save=save, fold=fold, fit_kwargs=fit_kwargs,
                       plot_kwargs=plot_kwargs, groups=groups, verbose=verbose, display_format=display_format)
        except Exception as e:
            print(f"Error plotting {plot_type} plot: {str(e)}")



In [14]:
# plot_all_model_plots(tuned_best_model, scale=1, save=True, fold=None, fit_kwargs=None, plot_kwargs=None, groups=None, verbose=True, display_format='streamlit')

## Results

Accuracy: About 93.2% of the predictions are correct. Not bad, but remember, accuracy can be misleading.

Precision: Around 94.0% of the instances the model predicted as positive are actually positive.

Recall: About 73.7% of all actual positive instances were captured. This model performed a bit better than in scikit-learn.

F1 Score: The harmonic mean of precision and recall is approximately 82.6%. This is a better metric when you need a balance between precision and recall.

Specificity: You correctly identified about 98.7% of all negative cases. That is a very good result, however that is the easy part as most people do not default on their loans.

In [16]:
final_model = finalize_model(tuned_best_model)

In [18]:
pull(final_model)

Model  Accuracy     AUC  Recall   Prec.      F1  \
0  Extreme Gradient Boosting    0.9322  0.9614  0.7365  0.9396  0.8258   

    Kappa     MCC  
0  0.7844  0.7934

In [21]:
save_model(final_model, 'final_model')

Transformation Pipeline and Model Successfully Saved


(Pipeline(memory=FastMemory(location=C:\Users\Ramon\AppData\Local\Temp\joblib),
          steps=[('numerical_imputer',
                  TransformerWrapper(exclude=None,
                                     include=['Id', 'Age', 'Income',
                                              'Emp_length', 'Amount', 'Rate',
                                              'Percent_income', 'Default',
                                              'Cred_length'],
                                     transformer=SimpleImputer(add_indicator=False,
                                                               copy=True,
                                                               fill_value=None,
                                                               keep_empty_features=False,
                                                               missing_values...
                                grow_policy=None, importance_type=None,
                                interaction_constraints=None, lea